gcp_variant_transforms/beam_io/vcf_estimate_io.py

```diff
@@ -0,0 +1,217 @@
+from apache_beam.io.filesystems import FileSystems
+from apache_beam.io.iobase import Read
+from apache_beam.transforms import PTransform
+
+
+class VcfEstimate(object):
+  """Container for estimation data about the VCF file."""
+
+  def __init__(self,
+               file_name=None,  # type: str
+               estimated_line_count=None,  # type: float
+               samples=None,  # type: List[str]
+               size_in_bytes=None  # type: int
+              ):
+    # type: (...) -> None
+    """Initializes a VcfEstimate object.
+
+    Object records file name as well as the following metadata:
+      estimated line count - extracted by reading the first data line after
+          the last header cloumn, calculating its size in bytes and diving
+          the total size of the file by the size of single line.
+      samples - contains the names of samples in the file. Extracted by reading
+          the last header line and dropping the first 8-9 common columns.
+      size_in_bytes - actual size of the file. Extracted from file's metadata.
+
+    Args:
+      file_name: name of file
+      estimated_line_count: estimated number of variants
+      samples: sample names in the file
+      size_in_bytes: size of the file
+    """
+    self.file_name = file_name
+    self.estimated_line_count = estimated_line_count
+    self.samples = samples
+    self.size_in_bytes = size_in_bytes
+
+  def __eq__(self, other):
+    return self.file_name == other.file_name
+
+  def __repr__(self):
+    return 'File Name: {}, Line Count: {}, Samples: {}, Size: {}'.format(
+        self.file_name,
+        self.estimated_line_count,
+        self.samples,
+        self.size_in_bytes
+    )
+
+
+class VcfEstimateSource(filebasedsource.FileBasedSource):
+  """A source for reading VCF file estimates."""
+
+  def __init__(self,
+               file_pattern,
+               compression_type=CompressionTypes.AUTO,
+               validate=True):
+    # type: (str, str, bool) -> None
+    super(VcfEstimateSource, self).__init__(file_pattern,
+                                            compression_type=compression_type,
+                                            validate=validate,
+                                            splittable=False)
+    self._compression_type = compression_type
+
+
+  def read_records(
+      self,
+      file_name,  # type: str
+      unused_range_tracker  # type: range_trackers.UnsplittableRangeTracker
+      ):
+    # type: (...) -> Iterable[VcfEstimate]
+    header_size = 0
+    with FileSystems.open(
+        file_name, compression_type=self._compression_type) as file_to_read:
+      header_line = file_to_read.readline()
+      # Read and skip all header lines starting with ##. Make sure to calculate
+      # their total size, to marginally better approximate the line count.
+      while (not header_line or
+             not header_line.strip() or header_line.startswith('##')):
+        if header_line and header_line.strip():
```

Comment:
We can remove this if condition to increase readability. Also, theoretically, the white spaces should also be considered in the size.